Enable --nv option to bind in nvidia drivers

ctmadison opened the pull request:
**Description of the Pull Request (PR):**

Add in support for automatically binding Nvidia libraries/drivers with a cmd line --nv argument.


**This fixes or addresses the following GitHub issues:**

- Fixes #1886


**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [ ] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [ ] I have tested this PR locally with a `make test`
- [x] This PR is against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge


Attn: @singularityware-admin

Commit:
Initial nvidia --nv development

Commit:
--nv dev work

ctmadison committed:
Merge branch 'master' of https://github.com/singularityware/singularity

ctmadison committed:
Merge branch 'master' into enable_nv

ctmadison committed:
nvidia --nv dev snapshot

ctmadison committed:
nvidia --nv dev_snapshot

ctmadison committed:
nvidia --nv cleanup

ctmadison committed:
nvidia --nv fix nvliblist.conf path

ctmadison committed:
update Gopkg

ctmadison committed:
Merge branch 'master' of https://github.com/singularityware/singularity

ctmadison committed:
update Gopkg

ctmadison committed:
Merge branch 'master' into enable_nv

ctmadison committed:
nvidia --nv bind so.1 files

ctmadison committed:
nvidia --nv minor optimization

ctmadison committed:
Merge branch 'master' of https://github.com/singularityware/singularity

ctmadison committed:
Merge branch 'master' into enable_nv

ctmadison committed:
added self to CONTRIBUTORS..md

bauerm97 commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+	if err == nil {
+		cmd := exec.Command(command, "list", "--binaries", "--ipcs", "--libraries")
+		out, err := cmd.Output()
+		if err == nil {
```
```
if err != nil {
    return nil, fmt.Errorf("error running nvidia-container-cli: %v", err)
}
```

GodloveD replied:
Same comment as above.  As much as I dislike it, I think all of this code actually does need to be wrapped in the if statement b/c we don't want to bail if it doesn't work.

ctmadison replied:
Hmmm... nvidia-container-cli is an optional component. Don't want to return if error.

bauerm97 replied:
Definitely want to error if we're unable to execute `nvidia-container-cli` if it does exist in `$PATH`

GodloveD replied:
No I disagree. `nvidia-container-cli` is still in a pretty early state afaik.  They have already changed he cli on me once.  If the user has a newer version of the binary installed that doesn't do what we expect we still need the option of falling back to the conf file instead.

bauerm97 commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+
+	// use nvidia-container-cli (if present)
+	command, err := exec.LookPath("nvidia-container-cli")
+	if err == nil {
```
```
if err != nil {
    return nil, fmt.Errorf("nvidia-container-cli not found: %v", err)
}
```

GodloveD replied:
That's what I was thinking at first too @bauerm97 but I don't think that will work in this instance.  If singularity cannot find `nvidia-container-cli` or if `nvidia-container-cli` returns an error it should just issue a warning and then move on to try to use the conf file instead.

ctmadison replied:
Hmmm... nvidia-container-cli is an optional component. Don't want to return if error.

bauerm97 replied:
Fair enough.

bauerm97 replied:
In this case, I would actually say to break these into functions:

```
func nvidiaContainerCli() ([]string, error) {
    // use nvidia-container-cli (if present)
    command, err := exec.LookPath("nvidia-container-cli")
    if err =! nil {
        return nil, nil
    }
}      
```

Then within the main code can simply do this:

```
s, err := nvidiaContainerCli()
if err != nil {
    return nil, fmt.Println("Unable to do something...: %v", err)
}
...
```

ctmadison replied:
Ok, I like this better...

bauerm97 commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+
+// GetNvidiaBindPath returns a string array consisting of filepaths of nvidia
+// related files to be added to the BindPaths
+func GetNvidiaBindPath(abspath string) []string {
```
`func GetNvidiaBindPath(abspath string) ([]string, error) {`

bauerm97 commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+	// grab the entries in nvliblist.conf file
+	// use ldconfig to pattern match from ld.so.cache
+	newpath, err := filepath.Glob(abspath + "/nvliblist.conf")
+	if err == nil {
```
```
if err != nil {
    return nil, fmt.Errorf("filepath.Glob() failed: %v", err)
}
```

GodloveD replied:
I agree with this though I think it should just spit a warning and then the function should return.

bauerm97 replied:
If you think it should spit a warning, then the person who calls this function can look at the error and determine if it should warn or error

ctmadison replied:
agreed

bauerm97 commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+		for _, filename := range newpath {
+
+			file, err := os.Open(filename)
+			if err == nil {
```
```
if err != nil {
    return nil, fmt.Errorf("failed to open file %v: %v", filename, err)
}
```

GodloveD replied:
I agree with this though I think it should just spit a warning and then the function should return.

bauerm97 commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+	// the nvliblist.conf file (ldconfig filenames are full filepaths)
+	var searchFileName string
+	command, err = exec.LookPath("ldconfig")
+	if err == nil {
```
```
if err != nil {
    return nil, fmt.Errorf("ldconfig not in path: %v", err)
}
```

GodloveD replied:
I agree with this though I think it should just spit a warning and then the function should return.

bauerm97 commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+	if err == nil {
+		cmd := exec.Command(command, "-p")
+		out, err := cmd.Output()
+		if err == nil {
```
```
if err != nil {
    return nil, fmt.Errorf("failed running ldconfig -p: %v", err)
}
```

GodloveD replied:
I agree with this though I think it should just spit a warning and then the function should return.

bauerm97 commented on src/cmd/singularity/cli/actions.go:
```diff
@@ -208,6 +209,11 @@ func execStarter(cobraCmd *cobra.Command, image string, args []string, name stri
 	}
 
+	if Nvidia {
+		NvidiaBindPaths := nvidiautils.GetNvidiaBindPath(buildcfg.SINGULARITY_CONFDIR)
```
```
NvidiaBindPaths, err := nvidiautils.GetNvidiaBindPath(buildcfg.SINGULARITY_CONFDIR)
if err != nil {
    sylog.Fatalf("Unable to capture nvidia bind points: %v", err)
}
```

GodloveD replied:
Maybe it should be a warning instead of a failure though?  That is the way 2.x works.

bauerm97 reviewed (changes requested):
This needs some readability tweaks before I'll be able to dive in further, once you fix what I've pointed out I'll take a second glance

GodloveD commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+	var strArray []string
+	var bindArray []string
+	var searchArray []string
+	var commentID = regexp.MustCompile(`#`)
```
I'm not expert on regular expressions, but this seems like it is saying "if there is a `#` anywhere in the line it's a comment".  

Shouldn't it be like "`^#`" or something?

ctmadison replied:
probably....

GodloveD commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+						fileName := fileNames[len(fileNames)-1]
+
+						testString := line + ":/.singularity.d/libs/" + fileName
+						bindArray = append(bindArray, testString)
```
appending to a string array in a loop sounds like it will be very slow to me.  I think there are faster ways to do this?  But I don't want us spending time on premature optimization if this is already pretty quick.

ctmadison replied:
I don't know another faster way (maybe other than building everything from scratch using buffers, etc.)

GodloveD commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+		}
+	}
+
+	cliEntries := strings.Join(strArray, " ") // save away for later comparison check (disallow duplicates)
```
Is it necessary to have the second variable `strArray` here?  Can `cliEntries` be constructed from `bindArray`?

ctmadison replied:
I don't think so. Because the bind array has entries like /dir/dir/file:/.singularity.d/file in it. I don't think it would be any faster and it would be more work. Are you thinking of a specific method?

GodloveD commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+
+	// grab the entries in nvliblist.conf file
+	// use ldconfig to pattern match from ld.so.cache
+	newpath, err := filepath.Glob(abspath + "/nvliblist.conf")
```
Oh so we are now using the _union_ of what we find in the conf file and what `nvidia-container-cli` yields?  I'm not sure I like that.  Kindof think we should default to `nvidia-container-cli` output and only fall back to conf file if it fails.

jmstover replied:
nvidia-container-cli didn't provide everything though right? I thought there was some stuff missing.

ctmadison replied:
Well, the nvidia-container-cli function only returns the /dir/file.so.xxx.xx entries (where the xxx.xx is the driver version). I can't just strip off the .xxx.xx and use the so entries because some things (like nvidia-smi binary look for the file.so.1 entry). I could make it work if that is the way you think I should go...

GodloveD replied:
I think what you want to do is use _either_ the output of `nvidia-container-cli` _or_ the output of the `nvidia.conf` file as a set of things to search the `ld.so.cache` for (through `ldconfig` output).  

Does that make sense?  In other words if you get usable output from `nvidia-container-cli` skip parsing the `nvidia.conf` file and just proceed directly to pattern matching between `nvidia-container-cli` output and `ldconfig` output.

GodloveD replied:
If you want to just append `.1` to all of the `nvidia-container-cli` entries and search for them as well, I think it's reasonable?  

@jmstover yes there was a recent issue where a user was confused b/c `nvidia-container-cli` does not provide open GL libraries, but that behavior is correct and our behavior is wrong.  We should comment those libs out of the conf file imho because those libs are not strictly tied to the version of the driver as the other libs are.

jmstover replied:
@GodloveD 
Maybe not on _recent_ drivers with libglvnd. But my `375.39` driver version, the GL libs *are* tied to the driver version.

i.e.  libGLX_nvidia.so.375.39

And we had cards where we couldn't update past a specific driver version because they removed support. So... no recent driver to have any of the new fancy stuff.

GodloveD replied:
@ctmadison I just had a brain wave.  I think you need to strip driver version from the `nvidia-container-cli` output and then use that as a _prefix_ when searching the output of `ldconfig`.  Tht way you will get any stray `.so.1` files (and other file appends) too.

ctmadison replied:
Yes, that should work. But, there is still the issue that @jmstover brought up about the GL libs that we have to agree on.

ArangoGutierrez reviewed (changes requested):
`make install` should cp `etc/nvliblist.conf`  to `$PREFIX/etc/nvliblist.conf`

GodloveD reviewed (changes requested):
I'm also not done looking at this yet, but in the interest of time I'm making a first round of comments.

GodloveD commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+	// use ldconfig to pattern match from ld.so.cache
+	newpath, err := filepath.Glob(abspath + "/nvliblist.conf")
+	if err == nil {
+		for _, filename := range newpath {
```
why is this in a for loop?  It should only be a single file right?

ctmadison replied:
yes, you are correct.

GodloveD commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+	if err == nil {
+		for _, filename := range newpath {
+
+			file, err := os.Open(filename)
```
I think you can use `defer` here to close when done?  Is that right?

ctmadison replied:
In my research, it was recommended to NOT use defer. I had it that way at first but changed it. But I will defer to others with more go experience...

GodloveD commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+				scanner := bufio.NewScanner(file)
+				for scanner.Scan() {
+					line := scanner.Text()
+					val := commentID.FindString(line)
```
For a possible simpler and more readable solution check this out (though I don't know if performance would suffer greatly).
https://golang.org/pkg/strings/#HasPrefix

ctmadison replied:
Yes, I started using more string. functions after I wrote this code and didn't take time to go back and change it.

GodloveD commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,134 @@
+				if ldconfigOutputline != "" {
+
+					for _, nvidiaConfFileline := range searchArray {
+						if nvidiaConfFileline != "" {
```
This should already be handled above right?  nvidiaConfFileline should not include empty entries afaik.

https://github.com/singularityware/singularity/blob/672f0e753500b55b26a04f736fb72b3c47ba3216/src/pkg/util/nvidiautils/nvidiautils.go#L36

ctmadison replied:
agreed.

ctmadison committed:
nvidia --nv addressed some PR comments

ctmadison committed:
nvidia --nv broke out nvidia-container-cli function

bauerm97 commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,142 @@
+	cmd := exec.Command(command, "list", "--binaries", "--ipcs", "--libraries")
+	out, err := cmd.Output()
+	if err != nil {
+		return nil, nil, nil
```
If we're unable to execute the `nvidia-container-cli` command, we *should* return an error:

```
if err != nil {
    return nil, nil, fmt.Errof("Unable to execute ldconfig: %v", err)
}
```

We should only be ignoring errors in the case where `nvidia-container-cli` doesn't exist in path

bauerm97 commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,142 @@
+
+	bindArray, strArray, err := nvidiaContainerCli()
+	if err != nil {
+		sylog.Warningf("no nvidia-container-cli returned: %v", err)
```
If we're only returning errors which *shouldn't* be ignored, then we should `sylog.Fataf(...)` instead.

bauerm97 commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,142 @@
+
+	// grab the entries in nvliblist.conf file
+	// use ldconfig to pattern match from ld.so.cache
+	file, err := os.Open(abspath + "/nvliblist.conf")
```
`defer file.Close()`

ctmadison committed:
nvidia --nv updated error returns

ctmadison committed:
nvidia --nv defer file close

ctmadison committed:
nvidia --nv refactor for using either nvidia-container-cli OR nvliblist.conf

GodloveD committed:
merge master and manually resolve conficts

GodloveD committed:
resolving paths for new repo location

GodloveD committed:
changes proposed by godlovedc to 3.0 --nv option

GodloveD commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,159 @@
+	for _, line := range strings.Split(string(out), "\n") {
+		if line != "" {
+
+			fileName := filepath.Base(line)
```
I'm not sure this will work when it gets input that does not include `/` or does include whitespace or whatever, but I cannot test b/c I don't have access to a GPU right now.

GodloveD commented on src/pkg/util/nvidiautils/nvidiautils.go:
```diff
@@ -0,0 +1,159 @@
+					ldconfigFileNames := strings.Split(ldconfigOutputSplitline[0], " ")
+					ldconfigFileName := strings.TrimSpace(string(ldconfigFileNames[0]))
+
+					if strings.HasPrefix(ldconfigFileName, nvidiaFileName) && ldconfigFileName != lastadd { // add if not duplicate
```
I think the addition of `strings.HasPrefix()` will pick up any `*.so.1` libs or others with different suffixes.  But needs to be tested to be sure.

GodloveD reviewed (dismissed):
I looked this over and added a few changes.  If they all work and are acceptable I approve.

jmstover commented on src/cmd/singularity/cli/actions.go:
```diff
@@ -206,6 +207,20 @@ func execStarter(cobraCmd *cobra.Command, image string, args []string, name stri
+		} else {
+			if len(BindPaths) == 0 {
+				sylog.Warningf("Could not find any NVIDIA libraries on this host!")
+				sylog.Warningf("You may need to edit %v/singularity/nvliblist.conf", buildcfg.SINGULARITY_CONFDIR)
```
This should be:  `%v/nvliblist.conf` ...   buldcfg.SINGULARITY_CONFDIR includes the singularity/ directory

jmstover commented on src/cmd/singularity/cli/actions.go:
```diff
@@ -206,6 +207,20 @@ func execStarter(cobraCmd *cobra.Command, image string, args []string, name stri
+		if err != nil {
+			sylog.Infof("Unable to capture nvidia bind points: %v", err)
+		} else {
+			if len(BindPaths) == 0 {
```
Should this be if `len(NvidiaBindPaths)` ?

GodloveD committed:
misnamed var

GodloveD committed:
fix path in warning

GodloveD committed:
check that conf file lines are NOT comments

ArangoGutierrez committed:
install nvliblist.conf on prefix on install

install $(SRC)/etc/nvliblist.conf to $(PREFIX)/etc/singularity/nvliblist.conf on make install

ArangoGutierrez committed:
fix some lint errors

ArangoGutierrez committed:
missing statement on Makefile.stub

ArangoGutierrez committed:
fix typo on line 21

ctmadison committed:
nvidia --nv centralize processing of binary files

jmstover reviewed (approved):
This is fully working for me now.

jmstover committed:
Merge branch 'master' into enable_nv

GodloveD reviewed (approved):
This looks good to me now.

bauerm97 committed:
Merge branch 'master' into enable_nv
